library/common/src/vsc_buffer.c

```diff
@@ -231,11 +225,27 @@ vsc_buffer_delete(vsc_buffer_t *self) {
         return;
     }
 
+    size_t old_counter = self->refcnt;
+    size_t new_counter = old_counter > 0 ? old_counter - 1 : old_counter;
+    #if defined(VSC_ATOMIC_COMPARE_EXCHANGE_WEAK)
+    //  CAS loop
+    while (!VSC_ATOMIC_COMPARE_EXCHANGE_WEAK(&self->refcnt, &old_counter, new_counter)) {
+        old_counter = self->refcnt;
+        new_counter = old_counter > 0 ? old_counter - 1 : old_counter;
+    }
+    #else
+    self->refcnt = new_counter;
+    #endif
+
+    if (new_counter > 0 || (new_counter == old_counter)) {
```

Comment:
Maybe new_counter == old_counter is better to replace with old_counter == 0. BTW, old_counter == 0 should not happen, maybe we should crash or something?

Reply:
`old_counter == 0` make sense because technically another thread can atomically modify `self->refcnt` and this thread can continue with non-atomic assignment `old_counter = self->refcnt;` within CAS loop.